zerver/lib/user_groups.py

```diff
@@ -125,6 +126,75 @@ def access_user_group_by_id(
     return user_group
 
 
+def access_user_group_for_deactivation(
+    user_group_id: int, user_profile: UserProfile
+) -> NamedUserGroup:
+    user_group = access_user_group_by_id(user_group_id, user_profile, for_read=False)
+
+    if (
+        user_group.direct_supergroups.exclude(named_user_group=None)
+        .filter(named_user_group__deactivated=False)
+        .exists()
+    ):
+        raise JsonableError(
+            _("You cannot deactivate a user group that is subgroup of any user group.")
+        )
+
+    anonymous_supergroup_ids = user_group.direct_supergroups.filter(
+        named_user_group=None
+    ).values_list("id", flat=True)
+
+    setting_group_ids_using_deactivating_user_group = [
+        *list(anonymous_supergroup_ids),
+        user_group.id,
+    ]
+
+    if Stream.objects.filter(
+        realm_id=user_group.realm_id,
+        deactivated=False,
+        can_remove_subscribers_group_id__in=setting_group_ids_using_deactivating_user_group,
+    ).exists():
+        raise JsonableError(_("You cannot deactivate a user group which is used for setting."))
+
+    if (
+        NamedUserGroup.objects.filter(realm_id=user_group.realm_id, deactivated=False)
+        .filter(
+            Q(can_manage_group_id__in=setting_group_ids_using_deactivating_user_group)
+            | Q(can_mention_group_id__in=setting_group_ids_using_deactivating_user_group)
+        )
```

Comment:
I think we're going to want to write this code to construct the `Q` objects using a loop over `GROUP_PERMISSION_SETTINGS`; that seems like the only mechanism that would be easy to keep correct as we add new settings.... both for this and the below similar blocks for realm/Stream.

Reply:
Updated.